File: DataStructures-Algorithms/11. Greedy/Coin_change_problem.c
/*
Given a value N, if we want to make change for N cents, and we have infinite supply of each of Coins = { C1, C2, .. , Cm} valued coins, 
what is the minimum number of coins to make the change?

Time Complexity : O(N).... 'N' is the value
Space Complexity : O(1) 
*/

#include<stdio.h>

void getMinimumCoins( int coins[], int size, int N)
{
	int minCoins=0;
	
	for( int i=size; i>=0; --i)
	{
		while( N >= coins[i])
		{
			N -= coins[i];
			
			printf(" %d,",coins[i]);
			
			++minCoins;
		}
	}
	
	printf("\n Min. Coins : %d ",minCoins);
}

void main()
{
	int N=93;
	
	int coins[] = { 1, 2, 5, 10, 20, 50, 100, 500, 1000 }; 
	
	int size = sizeof(coins)/sizeof(*coins);
	
	getMinimumCoins(coins,size-1,N);
}

File: interview/src/greedy/CoinChangeProblem.java
/**
 * Date 29/04/2022
 *
 * @author akhilpathivada
 *
 * https://www.geeksforgeeks.org/greedy-algorithm-to-find-minimum-number-of-coins/
 *
 * Time Complexity : O(N)
 * Space Complexity : O(1)
 *
 */
package greedy;

public class CoinChangeProblem {
        
        private int minCoins(int[] currency, int n) {
                if (n == 0) {
                        return 0;
                }
                int coins = 0;
                for (int i = currency.length - 1; i >= 0; --i) {
                        while (n >= currency[i]) {
                                n -= currency[i];
                                ++coins;
                        }
                }
                return coins;
        }
        
        public static void main(String[] args) {
                int[] currency = {1, 2, 5, 10, 20, 50, 100, 500, 1000};
                int n = 93;
                System.out.println(new CoinChangeProblem().minCoins(currency, n));
        }
}
